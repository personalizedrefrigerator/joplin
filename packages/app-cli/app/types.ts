export interface Gui {
	prompt(initialText: string, promptString?: string, options?: unknown): Promise<string>;
	stdoutMaxWidth(): number;
	isDummy(): boolean;

	showModalOverlay(text: string): void;
	hideModalOverlay(): void;
	maximizeConsole(): void;
	showConsole(): void;

	stdout(text: string): void;
	exit(): void;
}
