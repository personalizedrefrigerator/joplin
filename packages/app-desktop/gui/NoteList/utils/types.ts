import { FolderEntity, NoteEntity } from '@joplin/lib/services/database/types';
import { ListRenderer } from '@joplin/lib/services/plugins/api/noteListType';
import { PluginStates } from '@joplin/lib/services/plugins/reducer';
import { Size } from '@joplin/utils/types';
import { Dispatch } from 'redux';

export interface Props {
	themeId: any;
	selectedNoteIds: string[];
	notes: NoteEntity[];
	dispatch: Dispatch;
	watchedNoteFiles: string[];
	plugins: PluginStates;
	selectedFolderId: string;
	customCss: string;
	notesParentType: string;
	noteSortOrder: string;
	uncompletedTodosOnTop: boolean;
	showCompletedTodos: boolean;
	resizableLayoutEventEmitter: any;
	isInsertingNotes: boolean;
	folders: FolderEntity[];
	size: Size;
	searches: any[];
	selectedSearchId: string;
	highlightedWords: string[];
	provisionalNoteIds: string[];
	visible: boolean;
	focusedField: string;
	parentFolderIsReadOnly: boolean;
	listRenderer: ListRenderer;
}

export enum BaseBreakpoint {
	Sm = 75,
	Md = 80,
	Lg = 120,
	Xl = 474,
}

export interface Breakpoints {
	Sm: number;
	Md: number;
	Lg: number;
	Xl: number;
}
