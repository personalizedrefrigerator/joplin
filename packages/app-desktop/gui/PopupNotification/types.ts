import * as React from 'react';

export type PopupHandle = {
	remove(): void;
	scheduleDismiss(delay?: number): void;
};

export enum NotificationType {
	Info = 'info',
	Success = 'success',
	Error = 'error',
}

export type NotificationContentCallback = ()=> React.ReactNode;

export interface PopupOptions {
	type?: NotificationType;
}

export interface PopupControl {
	createPopup(content: NotificationContentCallback, props?: PopupOptions): PopupHandle;
}
