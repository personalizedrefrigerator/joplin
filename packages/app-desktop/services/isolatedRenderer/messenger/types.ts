
// Data that can be sent/received by a RemoteMessenger
export type SerializableData =
	number|boolean|string|undefined|null|SerializableData[]|{ readonly [key: string]: SerializableData };

// This provides some amount of type checking for messenger APIs.
//
// While all parameters and return types must also extend SerializableData,
// there's no clear way to enforce this in TypeScript (most methods lead to
// a "missing index signature" error).
export type ReturnsPromises<Type> = {
	[functionName in keyof Type]: (...args: any[])=> Promise<any>;
};

