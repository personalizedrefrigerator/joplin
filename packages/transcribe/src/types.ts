import type { Context } from 'koa';

declare module 'knex/types/tables' {
	interface Job {
		id: string;
		name: string;
		data: string;
		state: number;
		retry_count: number;
		output: string;
		started_on: string;
		completed_on: string;
		created_on: string;
		updated_on: string;
	}

	interface Queue {
		name: string;
		created_on: string;
		updated_on: string;
	}

	interface Tables {
		job: Job;
	}
}

export type Resource = {
	id: number;
	resource_path: string;
	created_time: Date;
	updated_time: Date;
};

export type JobData = {
	filePath: string;
};

export type Result = {
	result: string;
};

export interface BaseQueue {
	send(data: JobData): Promise<string>;
	fetch(): Promise<JobWithData | null>;
	fail(jobId: string, error: Error): Promise<void>;
	complete(jobId: string, data: Result): Promise<void>;
	getJobById(id: string): Promise<JobWithResult>;
	stop(): Promise<void>;
}

export interface ContentStorage {
	store(filepath: string): Promise<string>;
}

export type AppDefinedContext = {
	queue: BaseQueue;
	storage: ContentStorage;
};

export type AppContext = Context & AppDefinedContext;

export type JobWithData = {
	id: string;
	data: JobData;
};

export type OutputError = { stack: string; message: string };
export type OutputSuccess = { result: string };
export type Output = OutputError | OutputSuccess;

export type JobWithResult = {
	id: string;
	completedOn?: Date;
	output?: Output;
	state: string;
};

export enum JobStates {
	Created = 0,
	Retry = 1,
	Active = 2,
	Completed = 3,
	Cancelled = 4,
	Failed = 5,
}

export const jobStateToEnum = (j: JobStates) => {
	switch (j) {
	case 0:
		return 'created';
	case 1:
		return 'retry';
	case 2:
		return 'active';
	case 3:
		return 'completed';
	case 4:
		return 'cancelled';
	case 5:
		return 'failed';
	default:
		throw new Error(`Invalid job state: ${j}`);
	}
};

export interface WorkHandler {
	run(image: string): Promise<string>;
	init(): Promise<void>;
}

export type QueueConfiguration = {
	database: {
		name: string;
		user?: string;
		password?: string;
		port?: number;
	};
	ttl: number;
	retryCount: number;
	maintenanceInterval: number;
};
