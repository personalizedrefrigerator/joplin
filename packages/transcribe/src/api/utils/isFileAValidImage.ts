import { fromFile } from 'file-type';

export const supportedImageFormat = ['image/png', 'image/jpeg', 'image/bmp'];

const isFileAValidImage = async (filepath: string) => {
	const result = await fromFile(filepath);

	if (!result || !result.mime) {
		return [false, 'unknown'];
	}

	return [supportedImageFormat.includes(result.mime), result.mime];
};

export default isFileAValidImage;
