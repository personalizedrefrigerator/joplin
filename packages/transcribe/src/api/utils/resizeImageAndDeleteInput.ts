import { remove, copy } from 'fs-extra';
import * as sharp from 'sharp';

// Strange function that deletes the input file so we give it a specific name, because it shouldn't
// be used as a general function to resize an image. Should probably be refactored but for now it's
// good enough.
const resizeImageAndDeleteInput = async (inputPath: string, outputPath: string, imageMaxDimension: number) => {

	const metadata = await sharp(inputPath).metadata();

	if (!metadata || metadata.width === undefined || metadata.height === undefined) {
		await copy(inputPath, outputPath);
		await remove(inputPath);
		return;
	}

	if (Math.max(metadata?.width, metadata?.height) <= imageMaxDimension) {
		await copy(inputPath, outputPath);
		await remove(inputPath);
		return;
	}

	await sharp(inputPath)
		.resize(imageMaxDimension, imageMaxDimension, { fit: 'inside', withoutEnlargement: true })
		.toFile(outputPath);

	await remove(inputPath);
};

export default resizeImageAndDeleteInput;
