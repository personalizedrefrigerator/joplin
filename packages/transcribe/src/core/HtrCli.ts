import Logger from '@joplin/utils/Logger';
import { execCommand } from '@joplin/utils';
import { WorkHandler } from '../types';

const logger = Logger.create('HtrCli');

export default class HtrCli implements WorkHandler {

	private htrCliDockerImage: string;
	private htrCliImagesFolder: string;

	public constructor(htrCliDockerImage: string, htrCliImagesFolder: string) {
		this.htrCliDockerImage = htrCliDockerImage;
		this.htrCliImagesFolder = htrCliImagesFolder;
	}

	public async init() {
		logger.info('Loading');
		const result = await execCommand(`docker pull ${this.htrCliDockerImage}`, { quiet: true });
		logger.info('Finished loading: ', result);
	}

	public async run(imageName: string) {
		const command = `docker run --rm -t -v "${this.htrCliImagesFolder}:/images" ${this.htrCliDockerImage} ${imageName}`;

		logger.info('Running transcription...');
		logger.info(`Command: ${command}`);
		const result = await execCommand(command, { quiet: true });

		logger.info('Finished transcription');
		return this.cleanUpResult(result);
	}

	public cleanUpResult(transcriptionAndLogs: string) {
		const s1 = transcriptionAndLogs.split(/image decoded.*/);
		// Before the last `image decoded` line it is all logs generated by the transcription tool
		const everythingAfterImageDecoded = (s1[s1.length - 1]).trim();
		// After the transcription there are still some logs from the transcription tool
		const removedLastLogs = everythingAfterImageDecoded.slice(0, everythingAfterImageDecoded.indexOf('llama_perf_context_print:'));
		// Model is instructed to put transcription inside triple backticks
		return removedLastLogs.replace(/```/g, '').trim();
	}
}
