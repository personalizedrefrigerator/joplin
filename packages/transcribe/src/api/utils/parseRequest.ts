import { ErrorBadRequest } from '../../errors';
import { AppContext, JobData } from '../../types';
import isFileAValidImage, { supportedImageFormat } from './isFileAValidImage';

export const parseCreateJobRequest = async (ctx: AppContext) => {
	if (!ctx.request.files) throw new ErrorBadRequest('Invalid file property.');
	if (Array.isArray(ctx.request.files)) throw new ErrorBadRequest('Invalid file property.');
	if (!Object.keys(ctx.request.files).includes('file')) throw new ErrorBadRequest('Invalid file property.');
	if (Array.isArray(ctx.request.files.file)) throw new ErrorBadRequest('Invalid file property.');

	const file = ctx.request.files.file;

	if (!file) {
		throw new ErrorBadRequest('Request property "file" was not set.');
	}

	const [isValid, formatProvided] = await isFileAValidImage(file.filepath);

	if (!isValid) {
		throw new ErrorBadRequest(`Image format not accepted: ${formatProvided}. Try using: ${supportedImageFormat.join(' or ')}`);
	}

	return {
		storeImage: (file: string) => ctx.storage.store(file),
		sendToQueue: (data: JobData) => ctx.queue.send(data),
		filepath: file.filepath,
	};
};

export const parseGetJobRequest = (ctx: AppContext) => {
	const jobId = ctx.path.split('/')[ctx.path.split('/').length - 1];
	if (!jobId) {
		throw new ErrorBadRequest('Not possible to parse jobId value, expected: /transcribe/{job-uuid}');
	}

	return {
		jobId,
		getJobById: (jobId: string) => ctx.queue.getJobById(jobId),
	};
};
