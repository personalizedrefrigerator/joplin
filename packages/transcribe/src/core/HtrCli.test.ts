import { readFile } from 'fs-extra';
import HtrCli from './HtrCli';

describe('HtrCli', () => {
	const dt = new HtrCli('', '');
	it('should parse multiline result', async () => {
		const testCase = await readFile('./test-cases/1.txt');
		const result = dt.cleanUpResult(testCase.toString());
		expect(result).toMatchSnapshot();
	});
	it('should parse singleline result', async () => {
		const testCase = await readFile('./test-cases/2.txt');
		const result = dt.cleanUpResult(testCase.toString());
		expect(result).toMatchSnapshot();
	});
	it('should parse multiline result 2', async () => {
		const testCase = await readFile('./test-cases/3.txt');
		const result = dt.cleanUpResult(testCase.toString());
		expect(result).toMatchSnapshot();
	});
	it('should parse empty result', async () => {
		const testCase = await readFile('./test-cases/4.txt');
		const result = dt.cleanUpResult(testCase.toString());
		expect(result).toMatchSnapshot();
	});
	it('should parse empty result 2', async () => {
		const testCase = await readFile('./test-cases/5.txt');
		const result = dt.cleanUpResult(testCase.toString());
		expect(result).toMatchSnapshot();
	});
	it('should parse empty result 3', async () => {
		const testCase = await readFile('./test-cases/6.txt');
		const result = dt.cleanUpResult(testCase.toString());
		expect(result).toMatchSnapshot();
	});
});
