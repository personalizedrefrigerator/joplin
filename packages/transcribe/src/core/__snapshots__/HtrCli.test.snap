// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HtrCli should parse empty result 1`] = `"text"`;

exports[`HtrCli should parse empty result 2 1`] = `""`;

exports[`HtrCli should parse empty result 3 1`] = `"txt"`;

exports[`HtrCli should parse multiline result 1`] = `
"python
Kroken HTR

Tasks:
- Compare French HTR accuracy with Finetuned TROCR.
  - Set up comparison logic:
    - Kroken
    - TROCR
- Evaluate page segmentation performance. <--- T-C a seg
- Can there models run on end-user computers?
  - Kroken?
  - TROCR?"
`;

exports[`HtrCli should parse multiline result 2 1`] = `
"This is another mix of drawings and diagrams:

The above drawing is not text and should not be recognised as such.

This diagram has some text:

A
  \\  / 
   U B

This is more text.
This is even more."
`;

exports[`HtrCli should parse singleline result 1`] = `"This is a quick test of multi-line text."`;
